op-chain-ops/cmd/celo-migrate/db.go

```diff
@@ -34,35 +38,155 @@ func headerKey(number uint64, hash common.Hash) []byte {
 	return append(append(headerPrefix, encodeBlockNumber(number)...), hash.Bytes()...)
 }
 
-// openDB opens the chaindata database at the given path. Note this path is below the datadir
-func openDB(chaindataPath string) (ethdb.Database, error) {
+// Opens a database without access to AncientsDb
+func openDBWithoutFreezer(chaindataPath string, readOnly bool) (ethdb.Database, error) {
 	if _, err := os.Stat(chaindataPath); errors.Is(err, os.ErrNotExist) {
 		return nil, err
 	}
 
-	ldb, err := rawdb.Open(rawdb.OpenOptions{
-		Type:              "leveldb",
-		Directory:         chaindataPath,
-		AncientsDirectory: filepath.Join(chaindataPath, "ancient"),
-		Namespace:         "",
-		Cache:             DBCache,
-		Handles:           DBHandles,
-		ReadOnly:          false,
-	})
+	newDB, err := rawdb.NewLevelDBDatabase(chaindataPath, DBCache, DBHandles, "", readOnly)
 	if err != nil {
 		return nil, err
 	}
-	return ldb, nil
+
+	return newDB, nil
 }
 
-func createNewDbIfNotExists(newDBPath string) error {
+func createNewDbPathIfNotExists(newDBPath string) error {
 	if err := os.MkdirAll(newDBPath, 0755); err != nil {
 		return fmt.Errorf("failed to create new database directory: %w", err)
 	}
 	return nil
 }
 
+// readFullMigrationMarker reads a marker indicating whether a previous full migration attempt has modified the database
+func readFullMigrationMarker(db ethdb.KeyValueReader) ([]byte, error) {
+	return db.Get([]byte(CeloMigrationMarkerKey))
+}
+
+// writeFullMigrationMarker writes a marker indicating that a full migration has been attempted on the database
+func writeFullMigrationMarker(db ethdb.KeyValueWriter) error {
+	return db.Put([]byte(CeloMigrationMarkerKey), celoMigrationMarkerValue)
+}
+
+// checkForPrevFullMigration checks if a previous full migration has already been attempted on the database
+func checkForPrevFullMigration(newDBPath string, measureTime bool) (bool, error) {
+	if measureTime {
+		defer timer("checkForPrevFullMigration")()
+	}
+
+	newDB, err := openDBWithoutFreezer(newDBPath, true)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return false, nil
+		}
+		return false, fmt.Errorf("failed to open new database: %w", err)
+	}
+	defer newDB.Close()
+
+	marker, err := readFullMigrationMarker(newDB)
+	if err != nil {
+		if errors.Is(err, leveldb.ErrNotFound) {
+			return false, nil
+		}
+		return false, fmt.Errorf("failed to read celo migration marker: %w", err)
+	}
+
+	return len(marker) > 0, nil
+}
+
+// TODO(Alec) delete these
```

Comment:
Are these still to delete?

Reply:
Yea I'll do that, just wanted to make sure you were okay with the marker approach and didn't prefer one of these

Reply:
I couldn't get them to work on first pass but could give them another shot if we don't like the marker approach